LmDataset.py

```diff
@@ -1038,72 +1038,116 @@ class TranslationDataset(CachedDataset2):
                source_only=False,
                unknown_label=None,
+               source_factors=None,
+               target_factors=None,
+               factor_separator='|',
                seq_list_file=None,
                use_cache_manager=False,
                **kwargs):
     """
     :param str path: the directory containing the files
     :param str file_postfix: e.g. "train" or "dev". it will then search for "source." + postfix and "target." + postfix.
     :param bool random_shuffle_epoch1: if True, will also randomly shuffle epoch 1. see self.init_seq_order().
-    :param None|str source_postfix: will concat this at the end of the source. e.g.
-    :param None|str target_postfix: will concat this at the end of the target.
+    :param None|str source_postfix: will append this token to the source words (for all source factors).
+    :param None|str target_postfix: will append this token to the target words (for all target factors).
       You might want to add some sentence-end symbol.
     :param bool source_only: if targets are not available
-    :param str|None unknown_label: "UNK" or so. if not given, then will not replace unknowns but throw an error
+    :param str|dict[str,str]|None unknown_label: Label to replace out-of-vocabulary words with, e.g. "<UNK>".
+      If not given, will not replace unknowns but throw an error. Can also be a dict data_key -> unknown_label
+      to configure for each data key separately (default for each key is None).
+    :param list[str]|None source_factors: Names of source factors (excluding first factor, which is always
+      called 'data'). Words in source file have to have that many factors. Also, a vocabulary
+      "<factor_name>.voc.pkl" has to exist for each factor.
+    :param list[str]|None target_factors: analogous to source_factors. Excluding first factor, which is always
+      called 'classes'.
+    :param str factor_separator: string to separate factors of the words. E.g. if "|", words are expected to be
+      of format "<lemma>|<factor>|...".
     :param str seq_list_file: filename. line-separated list of line numbers defining fixed sequence order.
       multiple occurrences supported, thus allows for repeating examples while loading only once.
     :param bool use_cache_manager: uses :func:`Util.cf` for files
     """
 
     super(TranslationDataset, self).__init__(**kwargs)
+    assert os.path.isdir(path)
     self.path = path
     self.file_postfix = file_postfix
     self.seq_list = [int(n) for n in open(seq_list_file).read().splitlines()] if seq_list_file else None
-    if self._main_data_key is None:
-      self._main_data_key = "data"
-    if self._main_classes_key is None:
-      self._main_classes_key = "classes"
-    self._add_postfix = {self._main_data_key: source_postfix, self._main_classes_key: target_postfix}
-    self._keys_to_read = [self._main_data_key, self._main_classes_key]
+    self._add_postfix = {self.source_file_prefix: source_postfix.strip(),
```

Comment:
Add a newline after `{`. This would look nicer visually, as it then aligns with the next line.